examples/channel_inversion/inverse_problem.py

```diff
@@ -0,0 +1,161 @@
+from model_config import construct_solver
+
+solver_obj = construct_solver(
+    output_directory='outputs', store_station_time_series=False)
+options = solver_obj.options
+mesh2d = solver_obj.mesh2d
+bathymetry_2d = solver_obj.fields.bathymetry_2d
+manning_2d = solver_obj.fields.manning_2d
+elev_init_2d = solver_obj.fields.elev_init_2d
+
+# choose control variable(s)
+controls = [
+    # 'Bathymetry',
+    'Manning',
+    # 'InitialElev',
+]
+output_dir_suffix = '_' + '-'.join(controls) + '-opt'
+options.output_directory += output_dir_suffix
+
+gamma_hessian_list = []
+control_bounds_list = []
+op = inversion_tools.OptimisationProgress(options.output_directory)
+
+for control_name in controls:
+    if control_name == 'Bathymetry':
+        bathymetry_2d.assign(5.0)
+        bounds = [1.0, 50.]
+        op.add_control(bathymetry_2d)
+        gamma_hessian = Constant(1e-3)  # regularization parameter
+    elif control_name == 'Manning':
+        manning_2d.assign(1.0e-03)
+        bounds = [1e-4, 1e-1]
+        op.add_control(manning_2d)
+        gamma_hessian = Constant(1.0)
+    elif control_name == 'InitialElev':
+        elev_init_2d.assign(0.5)
+        bounds = [-10., 10.]
+        op.add_control(elev_init_2d)
+        gamma_hessian = Constant(0.1)
+    else:
+        raise ValueError(f'Unsupported control variable {control_name}')
+    print_output(f'{control_name} regularization params: hess={float(gamma_hessian):.3g}')
+    gamma_hessian_list.append(gamma_hessian)
+    control_bounds_list.append(bounds)
+# reshape to [[lo1, lo2, ...], [hi1, hi2, ...]]
+control_bounds = numpy.array(control_bounds_list).T
+
+print_output('Exporting to ' + options.output_directory)
+solver_obj.assign_initial_conditions(elev=elev_init_2d, uv=Constant((1e-5, 0)))
+
+# define the cost function
+# scale cost function so that J ~ O(1)
+dt_const = solver_obj.dt
+total_time_const = options.simulation_end_time
+J_scalar = dt_const/total_time_const
+
+observation_data_dir = 'outputs_forward'
+variable = 'elev'
+station_names = [
+    'stationA',
+    'stationB',
+    'stationC',
+    'stationD',
+    'stationE',
+]
+stationmanager = inversion_tools.StationObservationManager(
+    mesh2d, J_scalar=J_scalar, output_directory=options.output_directory)
+stationmanager.load_observation_data(observation_data_dir, station_names, variable)
+stationmanager.set_model_field(solver_obj.fields.elev_2d)
+
+# regularization for each control field
+reg_manager = inversion_tools.ControlRegularizationManager(
+    op.control_coeff_list, gamma_hessian_list, J_scalar=J_scalar)
+
+
+def cost_function():
+    """
+    Compute square misfit between data and observations.
+    """
+    t = solver_obj.simulation_time
+
+    J_misfit = stationmanager.eval_cost_function(t)
+    op.J += J_misfit
+
+
+def gradient_eval_callback(j, djdm, m):
+    """
+    Stash optimisation state.
+    """
+    op.set_control_state(j, djdm, m)
+    op.nb_grad_evals += 1
```

Comment:
We could get `op` to spit this out to avoid duplication in other scripts. I am happy to do that later.